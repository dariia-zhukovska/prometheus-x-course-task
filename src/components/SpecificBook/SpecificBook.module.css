.bookItem {
  display: flex;
  padding: 8px 25px;
}

.bookItemLeft {
  margin-right: 40px;
  height: 340px;
  box-shadow: 1px 1px 20px rgba(20, 20, 20, 0.25);
}

.bookItemRight,
.bookItemInfo {
  display: flex;
  flex-direction: column;
  gap: 8px;
  width: 80%;
}

.bookItemTitle {
  font-weight: 500;
  font-size: 4vh;
  color: #000000;
}

.bookItemAuthor {
  font-weight: 500;
  font-size: 24px;
  line-height: 29px;
  color: #797979;
}

.bookItemShortDescription,
.bookItemDescription {
  font-weight: 400;
  font-size: 16px;
  line-height: 20px;
  display: flex;
  align-items: center;
  color: #3a3a3c;
  line-height: 20px;
}

.bookItemMore {
  margin-bottom: 16px;
  font-weight: 700;
  font-size: 16px;
  background-size: 12px;
}

.bookItemPriceLine {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: max-content;
}

.bookItemPriceLine div {
  display: flex;
  gap: 2rem;
}

.bookItemPrice {
  align-self: center;
  width: max-content;
  font-weight: 700;
  font-size: 1.5rem;
  line-height: 34px;
  color: #f3564f;
}

.bookItemTotalPrice {
  align-self: center;
  color: #f45650;
  font-weight: 600;
  font-size: 22px;
}

.ButtonAddToCart {
  width: max-content;
  padding: 12px 40px;
  font-weight: 600;
  font-size: 20px;
  color: #fff;
  border-radius: 8px;
  border-style: none;
  background-color: #f45650;
  transition: all 0.2s ease-in-out;
}

.ButtonAddToCart:hover {
  transform: scale(1.1);
}

@media screen and (max-width: 480px) {
  .bookItem {
    display: flex;
    flex-direction: column;
  }

  .bookItemImage {
    margin: 16px 0;
    width: auto;
  }

  .bookItemRight {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin: 16px;
  }

  .bookItemPriceLine {
    justify-content: center;
    flex-wrap: wrap;
    gap: 1rem;
    width: 100%;
  }

  .ButtonAddToCart {
    margin-top: 16px;
    width: 100%;
  }
}

@media screen and (max-width: 767px) {
  .bookItem {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .bookItemImage {
    max-width: 300px;
    margin-bottom: 16px;
  }

  .bookItemRight {
    align-self: center;
    align-items: center;
    width: 100%;
  }

  .bookItemLeft {
    align-self: center;
    margin-right: 0;
    margin-bottom: 16px;
  }

  .bookItemInfo {
    margin-bottom: 16px;
    text-align: center;
  }

  .bookItemMoreCard {
    margin-top: 16px;
    text-align: center;
  }

  .bookItemPriceLine {
    flex-direction: column;
    align-items: center;
  }

  .bookItemPriceLine div:last-child {
    display: block;
    margin-right: 0;
  }


  .bookItemTotalPrice {
    margin-top: 8px;
  }

  .ButtonAddToCart {
    margin-top: 16px;
    width: 100%;
  }
}

@media (max-width: 1023px) {
  .bookItem {
    position: relative;
  }

  .bookItemPriceLine {
    display: flex;
    flex-direction: column;
  }

  .bookItemPriceLine div {
    margin-right: 0;
  }
}