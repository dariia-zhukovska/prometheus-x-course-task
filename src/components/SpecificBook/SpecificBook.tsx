import React, { useRef, useState, useEffect } from "react";
import { IBookListData, ICartBookData } from "../../types";
import styles from "./SpecificBook.module.css";
import imageNotFound from "../../assets/img/imageNotFound.png";
import { Link, useParams } from "react-router-dom";
import booksData from "../../data/books.json";
import Counter from "../Counter/Counter";
import PageNotFound from "../PageNotFound/PageNotFound";
import vectorUp from "../../assets/svg/icon_vector_up.svg";
import vectorDown from "../../assets/svg/icon_vector_down.svg";
import NavMenu from "../NavMenu/NavMenu";
import toast, { Toaster } from "react-hot-toast";
import useCart from "../../hooks/useCart";

export default function SpecificBook() {
  const dataBooks = booksData.books;
  const [isExpanded, setIsExpanded] = useState(true);
  const [countValue, setCountValue] = useState(1);
  const [totalPrice, setTotalPrice] = useState(1);
  const { cartItems, setCartItems } = useCart();
  console.log("BOOK", cartItems);

  const { id } = useParams();

  let selectedBook =
    dataBooks.find((book) => book.id === parseInt(`${id}`)) ||
    ({} as IBookListData);

  useEffect(() => {
    const currentItem = cartItems.find((book) => book.id === parseInt(`${id}`));
    if (currentItem) {
      setCountValue(currentItem.count);
      setTotalPrice(currentItem.totalPrice);
    }
  }, [cartItems, id]);

  useEffect(() => {
    const newTotalPrice = countValue * (selectedBook?.price || 0);
    setTotalPrice(newTotalPrice);
  }, [countValue, selectedBook?.price]);

  const handleToggle = () => {
    setIsExpanded(!isExpanded);
  };

  if (!selectedBook) {
    return <PageNotFound />;
  }

  const handleAddToCart = () => {
    const selectedCartItemIndex = cartItems.findIndex(
      (item: IBookListData) => item.id === selectedBook.id
    );

    if (selectedCartItemIndex !== -1) {
      const selectedCartItem = { ...cartItems[selectedCartItemIndex] };
      selectedCartItem.count = countValue;
      selectedCartItem.totalPrice = totalPrice;
      console.log(selectedCartItem);

      const newItems = cartItems.map((item) => {
        return +item.id === +selectedBook.id ? selectedCartItem : item;
      });
      setCartItems(newItems);
      toast.success("Cart has been updated successfully!");
    } else {
      const newItem = {
        id: selectedBook.id,
        image: selectedBook.image,
        title: selectedBook.title,
        author: selectedBook.author,
        price: selectedBook.price,
        count: countValue,
        totalPrice: totalPrice,
      };
      toast.success("Item added to the cart");
      setCartItems([...cartItems, newItem]);
    }
  };

  return (
    <>
      <Toaster />
      <NavMenu />
      <div className={styles.bookItem}>
        <div className={styles.bookItemLeft}>
          <img
            src={selectedBook.image || imageNotFound}
            alt="book_image"
            className={styles.bookItemImage}
          />
        </div>
        <div className={styles.bookItemRight}>
          <div className={styles.bookItemInfo}>
            <div className={styles.bookItemAuthor}>{selectedBook.author}</div>
            <div className={styles.bookItemTitle}>{selectedBook.title}</div>
            <div className={styles.bookItemShortDescription}>
              {isExpanded
                ? selectedBook.shortDescription
                : selectedBook.description}
            </div>
            <div className={styles.bookItemMoreCard}>
              <div
                className={styles.bookItemMore}
                onClick={handleToggle}
                style={{ backgroundColor: isExpanded ? vectorUp : vectorDown }}
              >
                {isExpanded ? "More about book" : "Less about book"}
              </div>
            </div>
          </div>
          <div className={styles.bookItemPriceLine}>
            <div>
              <div className={styles.bookItemPrice}>
                {`${selectedBook.price.toFixed(2)} $`}
              </div>
              <div className={styles.bookItemCount}>
                {<Counter count={countValue} setCount={setCountValue} />}
              </div>
            </div>
            <div>
              {" "}
              <div className={styles.bookItemTotalPrice}>
                {` Total Price: ${totalPrice.toFixed(2)}  $`}
              </div>
              <button
                className={styles.ButtonAddToCart}
                onClick={handleAddToCart}
              >
                Add To Cart
              </button>
            </div>
          </div>
        </div>
      </div>
    </>
  );
}
