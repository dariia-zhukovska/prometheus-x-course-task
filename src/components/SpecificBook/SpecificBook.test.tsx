
import { render, screen, act } from "@testing-library/react";
import SpecificBook from "./SpecificBook";

jest.mock("../NavMenu/NavMenu", () => () => null);

jest.mock("react-router-dom", () => ({
  useParams: () => ({
    id: 1,
  }),
}));

describe("testing decrease handler function", () => {
  test("should not call onCountChange, when count <= 1", async () => {
    const CountChangeMock = jest.fn();
    const { container } = render(<SpecificBook />);
    const decreaseButton = await screen.findByTestId("counter-decrease");
    expect(decreaseButton).toBeInTheDocument();
    const increaseButton = await screen.findByTestId("counter-increase");
    expect(increaseButton).toBeInTheDocument();
    const counterInput = await screen.findByTestId("counter-value");

    // @ts-ignore
    expect(counterInput.value).toEqual("1");
    await act(() => {
      increaseButton.click();
    });
    // @ts-ignore
    expect(counterInput.value).toEqual("2");
    await act(() => {
      decreaseButton.click();
    });
    // @ts-ignore
    expect(counterInput.value).toEqual("1");

    expect(container).toMatchSnapshot();
    // expect(CountChangeMock).not.toHaveBeenCalled();
  });
});
