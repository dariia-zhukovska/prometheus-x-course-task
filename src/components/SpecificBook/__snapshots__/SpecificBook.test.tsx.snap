// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing decrease handler function should not call onCountChange, when count <= 1 1`] = `
<div>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
  <div
    class="bookItem"
  >
    <div
      class="bookItemLeft"
    >
      <img
        alt="book_image"
        class="bookItemImage"
        src="https://courses.prometheus.org.ua/asset-v1:Ciklum+FEB101+2022_T3+type@asset+block@javascript_the_definitive_guide.jpg"
      />
    </div>
    <div
      class="bookItemRight"
    >
      <div
        class="bookItemInfo"
      >
        <div
          class="bookItemAuthor"
        >
          David Flanagan
        </div>
        <div
          class="bookItemTitle"
        >
          JavaScript: The Definitive Guide, 7th Edition
        </div>
        <div
          class="bookItemShortDescription"
        >
          JavaScript is the programming language of the web and is used by more software developers today than any other programming language.
        </div>
        <div
          class="bookItemMoreCard"
        >
          <div
            class="bookItemMore"
          >
            More about book
          </div>
        </div>
      </div>
      <div
        class="bookItemPriceLine"
      >
        <div>
          <div
            class="bookItemPrice"
          >
            10.99 $
          </div>
          <div
            class="bookItemCount"
          >
            <div
              class="counter"
            >
              <button
                data-testid="counter-decrease"
                style="color: grey;"
              >
                -
              </button>
              <input
                data-testid="counter-value"
                max="42"
                min="1"
                type="number"
                value="1"
              />
              <button
                data-testid="counter-increase"
              >
                +
              </button>
            </div>
          </div>
        </div>
        <div>
           
          <div
            class="bookItemTotalPrice"
          >
             Total Price: 10.99  $
          </div>
          <button
            class="ButtonAddToCart"
          >
            Add To Cart
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
